### ISSUE TITLE
Give backprop, grad, grad_zero to students

### ISSUE THREAD

**Issue by KonstiNik** [zincware/ZnPINNLab#9]
I would say mainly step 1 and 3 are relevant. Also we do not use weight decay in step 1. 

```python

def execute(
        model,
        train_ds: dict,
        test_ds: dict,
        lr = 0.01, 
        num_epochs = 300,
        activate_physics = False,
    ):
    """
    Execute the training procedure for a physics-informed neural network model.

    This function trains the model using specified hyperparameters and returns relevant data.
    The process involves the following steps:
    
    Step 1: Optimizer Initialization
            The optimizer is used to update the model's parameters during training. Therefore:
            - Initialize the optimizer, Adam with the model parameters 
              by calling the parameters() method on the model.
              (see [MODULE](https://pytorch.org/docs/stable/generated/torch.nn.Module.html))
            - Configure it with a specified learning rate 'lr'.
            More informations on the Adam optimizer can be found here:
            (see [ADAM](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html))
        
    Step 2: Lists Initialization
            In order to store the values of the models loss 
            and prediction during the training process, initialize:
            - The dictionary 'predictions_dict' 
            - The lists 'train_loss_evolution', 'test_loss_evolution' and 'predictions_list'. 


    Inside the training loop:

        Step 3: Loss Calculation
                Calculate the 'train_loss' and 'test_loss' 
                using the 'compute_total_loss' function.
                Make sure, 'activate_physics' is set to 'activate_physics'.

        Step 4: Training Step
                - Calculate the gradients of the model's parameters 
                  with respect to the total loss.
                  Do so by applying the 'backward' method on the train_loss. 
                  More information on this method here:
                  (see[backward](https://pytorch.org/docs/stable/generated/torch.Tensor.backward.html))
                - Then invoke the 'step' method on the optimizer 
                  in order to update the model parameters.
                  This Update is based on the calculated gradients beforehand. 
                  (see[step](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.step.html))
                - In preparation for the next epoch, 
                  reset the gradient of the optimizer to zero. 
                  Achieve this through by invoking the 'zero_grad' method on the optimizer.
                  (see[zero_grad](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html))

        Step 5: Loss Tracking
                For later analysis, append the current train and test loss values 
                to the appropriated lists.
                Make sure, the loss values are converted to float values 
                by applying the 'float()' class.
                (see[float](https://docs.python.org/3/library/functions.html#float)

        Step 6: Model Predictions Storage
                - Compute a model prediction using the inputs from 'test_ds'.
                - Store the current model predictions in the 'predictions_list'.

    Step 7: Dictionary Filling
            Fill and return the dictionary as specified below.

    Parameters
    ----------
    model : torch.nn.Module
        The neural network model to be trained and evaluated.
    train_ds : dict
        A dictionary containing training data, including inputs and targets.
    test_ds : dict
        A dictionary containing test data, including inputs and targets.
    lr : float, optional
        Learning rate for the optimizer. Default is 0.01.
    num_epochs : int, optional
        Number of training epochs. Default is 300.
    activate_physics : bool, optional
        Whether to activate the physics-informed loss. Default is False.

    Returns
    -------
    Dict
        A dictionary containing the following entries:
        - "train_loss_evolution": A list of training loss values.
        - "test_loss_evolution": A list of test loss values.
        - "predictions_list": A list of model predictions.
        - "train_targets_x": A tensor of x-coordinates of the training targets.
        - "train_targets_y": A tensor of y-coordinates of the training targets.
        - "test_targets_x": A tensor of x-coordinates of the test targets.
        - "test_targets_y": A tensor of y-coordinates of the test targets.
        if activate_physics:
        - "x_pred_col": A tensor of x-coordinates of the colocation points.
        - "y_pred_col": A tensor of y-coordinates of the colocation points.

    """    
    
    # Step 1: Initialize optimizer 
    optimizer = optim.Adam(model.parameters(), lr=lr) 
    
    # Step 2: Initialize the dictionary and lists
    predictions_dict = {}
    train_loss_evolution = []
    test_loss_evolution = []
    predictions_list = []
    
    # Define Loading Bar
    loading_bar = trange(1, num_epochs + 1)
    
    # Set up the training loop for the specified number of epochs ('num_epochs')
    for epoch in loading_bar:

        # Step 3: Compute the total loss for the training and the test data.
        train_loss = compute_total_loss(model, train_ds, activate_physics=activate_physics)
        test_loss = compute_total_loss(model, test_ds, activate_physics=activate_physics)
        
        # Step 4: Calculate the training loss ('train_loss') gradients using backpropagation, 
        #         then update the model parameters 
        #         and set the gradients to zero.
        train_loss.backward()
        optimizer.step()
        optimizer.zero_grad()

        # Step 5: Append the current loss value in the loss lists, initialized in Step 2.
        train_loss_evolution.append(float(train_loss))
        test_loss_evolution.append(float(test_loss))
           
        # Step 6: Compute the model prediction and store it in the respective list.
        predictions = model(test_ds['inputs'])
        predictions_list.append(predictions)

        # Print current epoch and losses
        loading_bar.set_description(f"Epoch: {epoch}")
        loading_bar.set_postfix({"Test Loss": test_loss.item(), "Train Loss": train_loss.item()})

    # Step 7: Fill the dictionary as specified and return it.
    predictions_dict["train_loss_evolution"] = train_loss_evolution
    predictions_dict["test_loss_evolution"] = test_loss_evolution
    predictions_dict["predictions_list"] = predictions_list
    predictions_dict["train_targets_x"] = train_ds["targets_x"]
    predictions_dict["train_targets_y"] = train_ds["targets_y"]
    predictions_dict["test_targets_x"] = test_ds["targets_x"]
    predictions_dict["test_targets_y"] = test_ds["targets_y"]
    
    if activate_physics:
        x_pred_col, y_pred_col = model(test_ds["t_col"])
        predictions_dict["x_pred_col"] = x_pred_col
        predictions_dict["y_pred_col"] = y_pred_col

    return predictions_dict

```
